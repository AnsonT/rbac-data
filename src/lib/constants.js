exports.ROOT_TENANT = '00000000-0000-0000-0000-000000000000'
exports.GLOBAL_TENANT = 'ffffffff-ffff-ffff-ffff-ffffffffffff'
exports.SUPERUSER_ROLE = '00000000-0000-0000-0000-000000000000'
exports.ADMIN_ROLE = '00000000-0000-0000-0000-000000000001'
exports.USER_ROLE = '00000000-0000-0000-0000-000000000002'
