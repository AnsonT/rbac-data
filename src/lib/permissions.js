import _ from 'lodash'
import uuid from 'uuid/v4'
import { InvalidParameterError } from './errors'
import { GLOBAL_TENANT, ROOT_TENANT } from './constants'
import { getUserByName } from './users'

function validatePermission (permission, global) {
  if (!permission) {
    return undefined
  }
  if (global && permission[0] !== '_') {
    throw new InvalidParameterError({ message: 'Global permissions must start with _' })
  }
  if (!global && permission[0] === '_') {
    throw new InvalidParameterError({ message: 'Local permissions cannot start with _' })
  }
  return permission
}

export async function createPermission (tx, { tenantId, permission, description, global, createdBy }) {
  const permissionId = uuid()
  if (permission.startsWith('_')) {
    global = true
  }
  if (tenantId && global) {
    throw new InvalidParameterError({ message: 'Cannot specify both tenantId and global: true' })
  }
  if (!global && !tenantId) {
    throw new InvalidParameterError({ message: 'Must specify tenantId, or set global: true' })
  }
  global = !tenantId
  tenantId = global ? GLOBAL_TENANT : tenantId
  permission = validatePermission(permission, global)
  const values = {
    permissionId,
    tenantId,
    permission,
    description,
    global,
    createdBy
  }
  await tx
    .insert(values)
    .into('permissions')
  return { permissionId, tenantId, permission, global }
}

export async function getPermissionById (tx, permissionId) {
  return tx
    .select()
    .where({ permissionId })
    .from('permissions')
    .first()
}

export async function getPermissionByName (tx, { tenantId, permission }) {
  if (tenantId) {
    return tx
      .select()
      .where({ permission })
      .where((builder) => builder.where({ tenantId }).orWhere({ tenantId: GLOBAL_TENANT }))
      .orderBy('global', 'asc')
      .from('permissions')
      .first()
  } else {
    return tx
      .select()
      .where({ permission })
      .where({ global: true })
      .from('permissions')
      .first()
  }
}

export async function listPermissions (tx, { tenantId = ROOT_TENANT } = {}) {
  return tx
    .select()
    .where({ tenantId })
    .orWhere({ tenantId: GLOBAL_TENANT })
    .orderBy('global', 'asc')
    .from('permissions')
}

export async function removePermission (tx, permissionId) {
  const count = await tx
    .from('permissions')
    .where({ permissionId })
    .whereNot({ canBeDeleted: false })
    .del()
  return { permissionId, count }
}

export async function updatePermission (tx, permissionId, { permission, description, modifiedBy }) {
  const { global } = await getPermissionById(tx, permissionId)
  permission = validatePermission(permission, global)
  const modifiedAt = tx.fn.now()
  const values = {
    permission,
    description,
    modifiedAt,
    modifiedBy
  }
  const count = await tx
    .where({ permissionId })
    .update(values)
    .into('permissions')
  return { count }
}

export async function listUserPermissions (tx, userId) {
  const permissions = await tx
    .select(['p.permissionId', 'p.permission', 'p.description', 'p.global', 'rp.denied'])
    .from('permissions as p')
    .join('rolesPermissions as rp', 'rp.permissionId', 'p.permissionId')
    .join('roles as r', 'r.roleId', 'rp.roleId')
    .join('usersRoles as ur', 'ur.roleId', 'r.roleId')
    .where('ur.userId', '=', userId)

  const allow = _.map(_.filter(permissions, { denied: false }), p => p.permission)
  const deny = _.map(_.filter(permissions, { denied: true }), p => p.permission)
  return {
    allow,
    deny,
    grants: _.filter(allow, a => !_.includes(deny, a)),
    permissions
  }
}

export function checkPermissions (requiredPermissions, grants, { any = false } = {}) {
  if (!grants || !_.isArray(grants)) {
    throw new InvalidParameterError({ message: 'grants must be an array' })
  }
  if (_.isString(requiredPermissions)) {
    return _.includes(grants, requiredPermissions)
  }
  if (_.isArray(requiredPermissions)) {
    const check = _.curry(checkPermissions)(_, grants)
    return any
      ? _.some(requiredPermissions, check)
      : _.every(requiredPermissions, check)
  }
  const allRequired = _.get(requiredPermissions, 'all')
  const anyRequired = _.get(requiredPermissions, 'any')
  if (allRequired && anyRequired) {
    throw new InvalidParameterError({ message: 'requirePermissions cannot have all and any in the same condition' })
  }
  if (allRequired) {
    return checkPermissions(allRequired, grants)
  }
  if (anyRequired) {
    return checkPermissions(anyRequired, grants, { any: true })
  }
  return false
}

export async function checkUserPermissionsByName (tx, requiredPermissions, { tenantId = ROOT_TENANT, userName }) {
  const { userId } = await getUserByName(tx, { tenantId, userName })
  const { grants } = await listUserPermissions(tx, userId)
  return checkPermissions(requiredPermissions, grants)
}

export async function listRolePermissionsByName (tx, { tenantId = ROOT_TENANT, roleName }) {
  return tx
    .select()
    .from('permissions as p')
    .join('rolesPermissions as rp', 'rp.permissionId', 'p.permissionId')
    .join('roles as r', 'r.roleId', 'rp.roleId')
    .where('r.roleName', '=', roleName)
    .where('r.tenantId', '=', tenantId)
}

export async function listRolePermissionsById (tx, roleId) {
  return tx
    .select()
    .from('permissions as p')
    .join('rolesPermissions as rp', 'rp.permissionId', 'p.permissionId')
    .join('roles as r', 'r.roleId', 'rp.roleId')
    .where('r.roleId', '=', roleId)
}
