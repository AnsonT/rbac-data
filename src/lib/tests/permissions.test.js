import _ from 'lodash'
import { UnitTestDb } from './db'
import { createPermissions, getPermissionById, getPermissionByName, listPermissions, removePermission, updatePermission } from '../permissions'
import { ROOT_TENANT, GLOBAL_TENANT } from '../constants'

const ACME_TENANT = '19443f78-47f2-4ed7-89eb-5707105ee51e'

const rootPerm1 = {
  permission: 'perm1',
  description: 'root permission 1',
  tenantId: ROOT_TENANT
}

const acmePerm1 = {
  permission: 'perm1',
  description: 'acme permission 1',
  tenantId: ACME_TENANT
}

const acme_Perm1 = { // eslint-disable-line camelcase
  permission: '_perm1',
  description: 'acme permission 1',
  tenantId: ACME_TENANT
}

const globalPerm2 = {
  permission: '_perm2',
  description: 'global permission 2',
  global: true
}

describe('Permissions Tests', () => {
  const db = new UnitTestDb()
  let knex = null

  beforeAll(async (done) => {
    await db.initialize('permissions')
    knex = db.knex
    done()
  })
  afterAll(async (done) => {
    await db.cleanup()
    done()
  })
  it('createPermissions local succeeds', async (done) => {
    const { permissionId } = await createPermissions(knex, rootPerm1)
    expect(permissionId).not.toBeNull()
    const { permission } = await getPermissionById(knex, permissionId)
    expect(permission).toBe(rootPerm1.permission)
    done()
  })
  it('createPermissions _ local permission fails', async (done) => {
    await expect(createPermissions(knex, acme_Perm1)).rejects.toThrow()
    done()
  })

  it('createPermissions duplicate fails', async (done) => {
    await expect(createPermissions(knex, rootPerm1)).rejects.toThrow()
    done()
  })

  it('createPermissions global succeeds', async (done) => {
    const { permissionId } = await createPermissions(knex, globalPerm2)
    expect(permissionId).not.toBeNull()
    const { permission } = await getPermissionById(knex, permissionId)
    expect(permission).toBe(globalPerm2.permission)
    done()
  })
  it('getPermissionByName global succeeds', async (done) => {
    const permission = await getPermissionByName(knex, { permission: globalPerm2.permission })
    expect(permission.permission).toBe(globalPerm2.permission)
    expect(permission.description).toBe(globalPerm2.description)
    expect(permission.global).toBe(globalPerm2.global)
    done()
  })
  it('getPermissionByName global with tentantId succeeds', async (done) => {
    const permission = await getPermissionByName(knex, { tenantId: ACME_TENANT, permission: globalPerm2.permission })
    expect(permission.permission).toBe(globalPerm2.permission)
    expect(permission.description).toBe(globalPerm2.description)
    expect(permission.global).toBe(globalPerm2.global)
    done()
  })

  it('getPermissionByName local fails without tenantId', async (done) => {
    const permission = await getPermissionByName(knex, { permission: rootPerm1.permission })
    expect(permission).toBeFalsy()
    done()
  })
  it('getPermissionByName local permissions succeeds', async (done) => {
    const permission = await getPermissionByName(knex, { tenantId: ROOT_TENANT, permission: rootPerm1.permission })
    expect(permission.permission).toBe(rootPerm1.permission)
    expect(permission.description).toBe(rootPerm1.description)
    expect(permission.global).toBe(false)
    done()
  })
  it('getPermissionByName global permissions with tenantId succeeds', async (done) => {
    const permission = await getPermissionByName(knex, { tenantId: ROOT_TENANT, permission: globalPerm2.permission })
    expect(permission.permission).toBe(globalPerm2.permission)
    expect(permission.description).toBe(globalPerm2.description)
    expect(permission.global).toBe(true)
    done()
  })
  it('createPermissions conflicted name in second tenant succeeds', async (done) => {
    const { permissionId } = await createPermissions(knex, acmePerm1)
    expect(permissionId).not.toBeNull()

    const { description } = await getPermissionById(knex, permissionId)
    expect(description).toBe(acmePerm1.description)
    done()
  })
  it('listPermissions succeeds', async (done) => {
    const omitMutableFields = (users) => _.map(users, user => _.omit(user, ['permissionId', 'createdAt', 'modifiedAt']))

    const rootPermissions = await listPermissions(knex)
    expect(omitMutableFields(rootPermissions)).toMatchSnapshot()
    const acmePermissions = await listPermissions(knex, { tenantId: ACME_TENANT })
    expect(omitMutableFields(acmePermissions)).toMatchSnapshot()
    const globalPermissions = await listPermissions(knex, { tenantId: GLOBAL_TENANT })
    expect(omitMutableFields(globalPermissions)).toMatchSnapshot()
    done()
  })
  it('updatePermissions succeeds', async (done) => {
    const { permissionId } = await getPermissionByName(knex, rootPerm1)
    const { count } = await updatePermission(knex, permissionId, { permission: 'newPerm1' })
    expect(count).toBe(1)
    const permission = await getPermissionById(knex, permissionId)
    expect(permission.permission).toBe('newPerm1')
    done()
  })

  it('removePermission succeeds', async (done) => {
    const { permissionId } = await getPermissionByName(knex, acmePerm1)
    const { count } = await removePermission(knex, permissionId)
    expect(count).toBe(1)
    const acmePermissions = await listPermissions(knex, { tenantId: ACME_TENANT })
    expect(acmePermissions.length).toMatchSnapshot()
    done()
  })
})
