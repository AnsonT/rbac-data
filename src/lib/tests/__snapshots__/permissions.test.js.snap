// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Permissions Tests listPermissions succeeds 1`] = `
Array [
  Object {
    "canBeDeleted": true,
    "description": "root permission 1",
    "global": false,
    "permission": "perm1",
    "tenantId": "00000000-0000-0000-0000-000000000000",
  },
  Object {
    "canBeDeleted": true,
    "description": "global permission 2",
    "global": true,
    "permission": "_perm2",
    "tenantId": "ffffffff-ffff-ffff-ffff-ffffffffffff",
  },
]
`;

exports[`Permissions Tests listPermissions succeeds 2`] = `
Array [
  Object {
    "canBeDeleted": true,
    "description": "acme permission 1",
    "global": false,
    "permission": "perm1",
    "tenantId": "19443f78-47f2-4ed7-89eb-5707105ee51e",
  },
  Object {
    "canBeDeleted": true,
    "description": "global permission 2",
    "global": true,
    "permission": "_perm2",
    "tenantId": "ffffffff-ffff-ffff-ffff-ffffffffffff",
  },
]
`;

exports[`Permissions Tests listPermissions succeeds 3`] = `
Array [
  Object {
    "canBeDeleted": true,
    "description": "global permission 2",
    "global": true,
    "permission": "_perm2",
    "tenantId": "ffffffff-ffff-ffff-ffff-ffffffffffff",
  },
]
`;

exports[`Permissions Tests removePermission succeeds 1`] = `1`;
